pupil_src/shared_modules/eye_movement_detector.py

```diff
@@ -0,0 +1,937 @@
+from tasklib.background.patches import Patch, IPCLoggingPatch
+from tasklib.manager import PluginTaskManager
+import file_methods as fm
+import player_methods as pm
+
+import methods
+from plugin import Analysis_Plugin_Base
+from observable import Observable
+import video_capture as vc
+
+# third-party
+import nslr_hmm
+import numpy as np
+import cv2
+from pyglui import ui
+from pyglui.cygl.utils import RGBA, draw_circle
+from pyglui.pyfontstash import fontstash
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: This protocol definition can be moved into `camera_models.py`
+class Camera_Model(Protocol):
+    def undistort(self, img: np.ndarray) -> np.ndarray:
+        ...
+
+    def unprojectPoints(
+        self, pts_2d: np.ndarray, use_distortion: bool = True, normalize: bool = False
+    ) -> np.ndarray:
+        ...
+
+    def projectPoints(
+        self,
+        object_points,
+        rvec: typing.Optional[np.ndarray] = None,
+        tvec: typing.Optional[np.ndarray] = None,
+        use_distortion: bool = True,
+    ):
+        ...
+
+    # def solvePnP(self, uv3d, xy): ...
+    def save(self, directory: str, custom_name: typing.Optional[str] = None):
+        ...
+
+
+class Immutable_Capture:
+    def __init__(self, capture: typing.Type[vc.base_backend.Base_Source]):
+        self.frame_size: typing.Tuple[int, int] = (
+            int(capture.frame_size[0]),
+            int(capture.frame_size[1]),
+        )
+        self.intrinsics: Camera_Model = capture.intrinsics
+        self.timestamps: np.ndarray = capture.timestamps
+
+
+Color_RGBA = typing.Tuple[float, float, float, float]
+
+
+@enum.unique
+class Segment_Method(enum.Enum):
+    GAZE = "gaze"
+    PUPIL = "pupil"
+
+
+@enum.unique
+class Segment_Class(enum.Enum):
+    FIXATION = "fixation"
+    SACCADE = "saccade"
+    POST_SACCADIC_OSCILLATIONS = "pso"
+    SMOOTH_PURSUIT = "smooth_pursuit"
+
+    @classmethod
+    def from_nslr_class(cls, nslr_class):
+        return {
+            nslr_hmm.FIXATION: Segment_Class.FIXATION,
+            nslr_hmm.SACCADE: Segment_Class.SACCADE,
+            nslr_hmm.PSO: Segment_Class.POST_SACCADIC_OSCILLATIONS,
+            nslr_hmm.SMOOTH_PURSUIT: Segment_Class.SMOOTH_PURSUIT,
+        }[nslr_class]
+
+    @property
+    def color_rgba(self) -> Color_RGBA:
+        # https://flatuicolors.com/palette/defo
+        grey = (52 / 255, 73 / 255, 94 / 255, 1.0)  # wet asphalt
+        yellow = (241 / 255, 196 / 255, 15 / 255, 1.0)  # sun flower
+        green = (39 / 255, 174 / 255, 96 / 255, 1.0)  # nephritis
+        blue = (41 / 255, 128 / 255, 185 / 255, 1.0)  # belize hole
+        purple = ((142 / 255, 68 / 255, 173 / 255, 1.0),)  # wisteria
+        return {
+            Segment_Class.FIXATION: yellow,
+            Segment_Class.SACCADE: green,
+            Segment_Class.POST_SACCADIC_OSCILLATIONS: blue,
+            Segment_Class.SMOOTH_PURSUIT: purple,
+        }.get(self, grey)
```

Comment:
Same argument as above to make this dict a class attribute. Also, I would not catch the case where `self` is not one of the possible values.